File: facade/src/main/java/org/tdf/sunflower/vm/hosts/U256Host.java
package org.tdf.sunflower.vm.hosts;

import org.tdf.evm.MutableBigInteger;
import org.tdf.evm.SlotUtils;
import org.tdf.lotusvm.runtime.HostFunction;
import org.tdf.lotusvm.types.FunctionType;
import org.tdf.lotusvm.types.ValueType;


import java.util.Arrays;
import java.util.Collections;


import static org.tdf.evm.SlotUtils.*;
import static org.tdf.sunflower.vm.WBI.*;

public class U256Host extends HostFunction {
    private static final int[] ZERO_SLOT = new int[SLOT_SIZE * 2];

    private final int[] slot0 = new int[SLOT_SIZE];
    private final int[] slot1 = new int[SLOT_SIZE];
    private final int[] varSlot0 = new int[SLOT_SIZE * 2];
    private final int[] varSlot1 = new int[SLOT_SIZE * 2];

    private final int[] divisor = new int[SLOT_SIZE * 2];


    private final byte[] tempBytes = new byte[SLOT_BYTE_ARRAY_SIZE];

    private final MutableBigInteger mut0 = new MutableBigInteger(slot0);
    private final MutableBigInteger mut1 = new MutableBigInteger(slot1);

    private final MutableBigInteger mut2 = new MutableBigInteger(varSlot0);
    private final MutableBigInteger mut3 = new MutableBigInteger(varSlot1);


    enum U256OP {
        ADD,
        SUB,
        MUL,
        DIV,
        MOD
    }

    public static final FunctionType FUNCTION_TYPE = new FunctionType(
        // offset, length, offset
        Arrays.asList(
            ValueType.I64, ValueType.I64, ValueType.I64,
            ValueType.I64, ValueType.I64, ValueType.I64,
            ValueType.I64, ValueType.I64, ValueType.I64
        ),
        Collections.singletonList(ValueType.I64));

    public U256Host() {
        super("_u256", FUNCTION_TYPE);
    }

    // override slot by uint256 pointer
    private void copyToSlot(int[] slot, long[] args, int offset) {
        // call peek function, convert uint256 pointer to byte array fat pointer
        for(int i = 0; i < 4; i++) {
            slot[i * 2] = (int) (args[offset + i] >>> 32);
            slot[i * 2 + 1] = (int) args[offset + i];
        }
    }

    // write slot into memory
    private long writeSlot(int[] slot) {
        long[] data = new long[slot.length / 2];
        for(int i = 0; i < data.length; i++) {
            data[i] = ((slot[i * 2] & LONG_MASK) << 32) | ((slot[i * 2 + 1] & LONG_MASK));
        }
        return INSTANCE.mallocWords(getInstance(), data);
    }

    @Override
    public long execute(long[] longs) {
        System.out.println("u256..");
        int i = (int) longs[0];

        U256OP op = U256OP.values()[i];



        copyToSlot(slot0, longs, 1);
        copyToSlot(slot1, longs, 5);

        if (op == U256OP.MUL || op == U256OP.DIV || op == U256OP.MOD) {
            mut0.setValue(slot0, SLOT_SIZE);
            mut0.normalize();
            mut1.setValue(slot1, SLOT_SIZE);
            mut1.normalize();
            mut2.clear();
        }


        switch (op) {
            case ADD: {
                Arrays.fill(varSlot0, 0);
                long carry = SlotUtils.add(slot0, 0, slot1, 0, varSlot0, 8);
                varSlot0[7] = (int) carry;
                return writeSlot(varSlot0);
            }
            case SUB: {
                SlotUtils.sub(slot0, 0, slot1, 0, slot0, 0);
                return writeSlot(slot0);
            }
            case MUL: {
                if (mut0.isZero() || mut1.isZero()) {
                    return writeSlot(ZERO_SLOT);
                }
                mut0.multiply(mut1, mut2);
                Arrays.fill(varSlot1, 0);
                System.arraycopy(varSlot0, mut2.getOffset(), varSlot1, SLOT_SIZE * 2 - mut2.getIntLen(), mut2.getIntLen());
                return writeSlot(varSlot1);
            }
            case DIV: {
                // reset temporary divisor and rem
                Arrays.fill(divisor, 0);
                mut3.clear();

                mut0.divideKnuth(mut1, mut2, mut3, divisor, false);

                Arrays.fill(slot0, 0);
                SlotUtils.trim256(mut2.getValue(), mut2.getOffset(), mut2.getIntLen(), slot0, 0);
                return writeSlot(slot0);
            }
            case MOD: {
                // reset temporary divisor and rem
                Arrays.fill(divisor, 0);
                mut3.clear();

                mut0.divideKnuth(mut1, mut2, mut3, divisor, true);

                Arrays.fill(slot0, 0);
                SlotUtils.trim256(mut3.getValue(), mut3.getOffset(), mut3.getIntLen(), slot0, 0);
                return writeSlot(slot0);
            }
            default:
                throw new UnsupportedOperationException();
        }
    }
}

File: facade/src/main/java/org/tdf/sunflower/vm/hosts/U256Host.kt
package org.tdf.sunflower.vm.hosts

import com.github.salpadding.evm.Slot
import com.github.salpadding.evm.SlotImpl
import org.tdf.lotusvm.runtime.HostFunction
import org.tdf.lotusvm.types.FunctionType
import org.tdf.lotusvm.types.ValueType
import org.tdf.sunflower.vm.WBI
import org.tdf.sunflower.vm.abi.WbiType

class U256Host : HostFunction("_u256", FUNCTION_TYPE) {
    private val slot0: Slot = SlotImpl()
    private val slot1: Slot = SlotImpl()
    private val slot2: Slot = SlotImpl()
    private val slot3: Slot = SlotImpl()

    private val tmpData = ByteArray(32)

    internal enum class U256OP {
        ADD, SUB, MUL, DIV, MOD
    }

    private fun clearTmp() {
        for(i in 0 until tmpData.size)
            tmpData[i] = 0
    }

    private fun loadSlot(offset: Int, slot: Slot) {
        val startAndLen = instance.execute(WBI.WBI_PEEK, offset.toLong(), WbiType.UINT_256)[0]
        val start = (startAndLen ushr 32).toInt()
        val len = startAndLen.toInt()

        clearTmp()
        for(i in 0 until len) {
            tmpData[tmpData.size - len + i] = instance.memory.load8(start + i)
        }

        slot.decodeBE(tmpData, 0)
    }

    private fun putSlot(slot: Slot): Long {
        clearTmp()
        slot.encodeBE(tmpData, 0)
        var firstNoZero = -1

        for(i in 0 until tmpData.size) {
            if(tmpData[i] != (0).toByte()) {
                firstNoZero = i
                break
            }
        }

        val size = if (firstNoZero < 0) { 0 } else { tmpData.size - firstNoZero }
        val ptr = (instance.execute(WBI.WBI_MALLOC, size.toLong())[0]).toInt()
        if (ptr < 0) throw RuntimeException("malloc failed: pointer is negative")

        for(i in 0 until size) {
            instance.memory.storeI8(ptr + i, tmpData[firstNoZero + i])
        }

        val p = instance.execute(WBI.WBI_CHANGE_TYPE, WbiType.UINT_256, ptr.toLong(), size.toLong())[0]
        val r = p.toInt()
        if (r < 0) throw RuntimeException("malloc failed: pointer is negative")
        return r.toLong()
    }

    override fun execute(args: LongArray): Long {
        val i = args[0].toInt()
        val op = U256OP.values()[i]
        loadSlot(args[1].toInt(), slot0)
        loadSlot(args[2].toInt(), slot1)

        when (op) {
            U256OP.ADD -> {
                slot0.add(slot1)
                return putSlot(slot0)
            }
            U256OP.SUB -> {
                slot0.sub(slot1)
                return putSlot(slot0)
            }
            U256OP.MUL -> {
                slot2.initZeros()
                slot0.mul(slot1, slot2)
                return putSlot(slot2)
            }
            U256OP.DIV -> {
                slot2.initZeros()
                slot3.initZeros()
                slot0.divMod(slot1, slot2, slot3)
                return putSlot(slot2)
            }
            U256OP.MOD -> {
                slot2.initZeros()
                slot3.initZeros()
                slot0.divMod(slot1, slot2, slot3)
                return putSlot(slot0)
            }
        }
    }

    companion object {
        val FUNCTION_TYPE = FunctionType( // offset, length, offset
            listOf(ValueType.I64, ValueType.I64, ValueType.I64), listOf(ValueType.I64)
        )
    }
}